adrs/5541831634/README.md

```diff
@@ -0,0 +1,160 @@
+* You **must** use external service brand name as a subdirectory in
+  `<fractalDir>/services/[service name]`.
+* You **must** use `<fractalDir>/components` to store reusable React
+  components. You may create a directory to scope the fractal directories to
+  a component, otherwise use use a file.
+* You **must** use `<fractalDir>/hooks` to store reusable React
+  hooks.
+* You **must** use `<fractalDir>/types` to store reusable TypeScript types.
+* You **must** use `<fractalDir>/styles` to store components styles such as
+  CSS or Stylesheets.
+* You **must** use `<fractalDir>/constants` to store constants variables.
+* You **must** use `<fractalDir>/contexts` to store React contexts definitions.
+* You **must** use `<fractalDir>/i18n` to store anything
+  related to Internationalization.
+* You **must** use `<fractalDir>/i18n/datetime` to store i18n date
+  and time formatting.
+* You **must** use `<fractalDir>/i18n/list` to store i18n sensitive
+  list formatting. Read more about the intent at [Intl.ListFormat](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/ListFormat).
+* You **must** use `<fractalDir>/i18n/numbers` to store i18n number
+  formatting. Read more about the intent at [Intl.NumberFormat](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/NumberFormat).
+* You **must** use `<fractalDir>/i18n/plural` to store i18n
+  pluralization formatting. Read more about the intent at [Intl.PluralRules](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/PluralRules).
+* You **must** use `<fractalDir>/i18n/relative-time` to store i18n
+  relative-time formatting. Read more about the intent at [Intl.RelativeTimeFormat](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/RelativeTimeFormat).
+* You **must** use `<fractalDir>/helpers/datetime` to store anything related
+  to Dates or Times. Except formatters, use `<fractalDir>/i18n/datetime`
+  instead.
+* You **must** use `<fractalDir>/helpers/currency` to store anything related
+  to currency. Except formatters, use `<fractalDir>/i18n/numbers`
+  instead.
+* You **must** use `<fractalDir>/helpers` to store anything that does not
+  fall into any of the other helper directory categories. You **may** create
+  your own files and directories based on your context.
+* You **must** use `<fractalDir>/transports/http` to store anything
+  related to HTTP transport clients.
+* You **must** use `<fractalDir>/transports/web-socket` to store
+  anything related to Web Socket transport clients.
+* You **must** use `<fractalDir>/transports/graphql` to store
+  anything related to GraphQL transport clients.
+* You **must** use `<fractalDir>/storages` to store anything related to the
+  device. You **may** use `local`, `session`, `cookies`, `websql`, and
+  `indexeddb` as subdirectories.
+* You **must** use `<fractalDir>/schemas` to store anything related to Joi, Yup
```

Comment:
Part of me is wondering if this ADR reaches too far.
It seems to try to cover every possible usecase for a complex NextJS project.

Has this entire structure (ie: every subfolder listed here) actually been implemented in a project before?
If it has been implemented in a project before what was the outcome? 
How many iterations have been made on this structure?

I'm concerned that if this structure is being imposed without actually using it first we may inadvertently introduce complexity.

Reply:
The answer is yes, it has been used before.

If it is common nowadays to have validations for your forms using Yup, Joi, or some similar library, for example:

- https://react-hook-form.com/get-started#SchemaValidation
- https://formik.org/docs/guides/validation#validationschema

Those two packages are the most popular at the moment in the ecosystem for dealing with form.

As you noticed they called it schemas because those packages themselves call themselves "schema" for validating data. That is the name comes from.

The directory name is closer to what the terminology is being used for today.

Reply:
I'm referring to all of the subfolders in the `fractal directory`
types, styles, contexts, transports, storages, schemas etc

Reply:
It depends, the intention of having directories for a particular group of things, yes, it is widely used.

1. Routing
2. Reusable components
3. Internationalization
  3.1 Numbers
  3.2 Dates
  4.3 Numbering things
  4.4 ...
4. Validations
5. Things that are hard to give a group (helpers)
6. Transport protocols
  6.1 HTTP
  6.2 GraphQL
  6.3 Socket
7. Dealing with devices storage

And many other things are commonly needed in apps, it may be the case that you wouldn't need all of them, but in case you do, you have a common ground.

This is leaving directories based on technology, which we could document as part of another ADR.